### ISSUE TITLE
jqGrid (v4.8.2 or free fork latest buid) and Select2 v4.0.0 on edit modal window

### ISSUE THREAD

**Issue by Wooya** [free-jqgrid/jqGrid#59]
I've described an issue on Select2 GitHub page, but they rejected it as not a bug. Maybe Oleg can help with this issue :-)
https://github.com/select2/select2/issues/3352

**Comment by OlegKi** [comment]
Could you provide the demo which reproduces the problem (as jsfiddle for example) with the test case (the descriptions of the steps to reproduce the problem, expected results and the results which one have instead)? Having the working demo one can reproduce the problem in the exact case which you have. One could debug the code and so one could localize the origin of the problem and finally to solve the problem.

**Comment by Wooya** [comment]
Here's jsfiddle, but... it's working :-( Try to type something in select2 input

https://jsfiddle.net/191no21j/

**Comment by OlegKi** [comment]
Is the problem is already solved for you or you need my help?

**Comment by Wooya** [comment]
I think it's solved, but question is about this behaviour. Why modal = true breaks Select2 functionality inside edit/add jqGrid windows?

**Comment by OlegKi** [comment]
The demo have some problems. First of all you miss `formatter: "select"` probably. Then you use wrong format for `editoptions.value`. It should be `value: { 1: "Poznan", 2: "Krakow", 3: "Warszawa", 4: "Berlin" }` or just `value: "1:Poznan;2:Krakow,3:Warszawa,4:Berlin"`. It's two formats which are supported in jqGrid 4.7. Free jqGrid have no extensions of the format. The format `value: "1:Poznan;2:Krakow,3:Warszawa,4:Berlin"` is recommended because it produce the same order of options in all web browsers.

The main problem which you describe can be solved you implementing modal dialog using jQuery UI instead of `jqModal.js` used by default.
```JavaScript
$.fn.jqm = false;
```
The alternative to the way is possible but it is really tricky. The code of original `jqModal.js` is very difficult to read. Because of that I makes a lot of renaming of internal variables and reformatted it in free jqGrid. The problem makes [the line](https://github.com/free-jqgrid/jqGrid/blob/master/js/jqmodal.js#L31) which will be used to determine which input field or which clicks of mouse are from "own" modal dialog and which one need be blocked. `jqmodal.js` uses the name of the class `".jqmID" + number` to identify the "own" controls. So one could add the same class to the container of select2. The code could be something like the following
````JavaScript
dataInit: function (elem) {
	$(elem).select2()
	.on("select2:open", function(event) {
		var i, name,
		dialogClasses = $(elem).closest(".ui-jqdialog")[0]
			.className.split(" ");
		for (i = 0; i < dialogClasses.length; i++) {
			name = dialogClasses[i];
			if (name.substring(0, 5) === "jqmID") { // jqmID1
				$("body>.select2-container").addClass(name);
			}
		}
	});
}
````
See https://jsfiddle.net/OlegKi/191no21j/10/

**Comment by Wooya** [comment]
Thanks a lot for explanation. I just follow known problems with select2 and jQueryUI modals. As usual your solution looks to solve the problem :-) Thanks again.

As far as I see I don't need to set modal = true, so easest way to solve this problem at the moment is modal = false, what I'm doing right now. This issue can be closed, I think.

**Comment by OlegKi** [comment]
OK. You are welcome!

**Comment by OlegKi** [comment]
Hi once more,

I thought about the problem and have found even better solution. I remind me that one had close problem with processing of mouse events in `jqModal.js` and I posted the fix many years ago which are included already in old version of jqGrid. The fix consist from testing **absolute** position of mouse cursor and if it was inside of the area of modal dialog then the mouse event was allowed.

So I had the idea to modify `jqModal.js` for making close tests for keyboard events too. I posted the corresponding changes to free jqGrid (see [here](https://github.com/free-jqgrid/jqGrid/commit/5267bc5a00d63d045e6c5405f19dc30061ed8926)) and now select2 have no problems. See http://jsfiddle.net/191no21j/13/.
